### PR TITLE
Add output dump for correctness of benchmark_app

### PR DESCRIPTION
**Pull request by xufang-lisa**
### Details:
 - *Add output dump for correctness of benchmark_app*

### Tickets:
 - *89166*

### CODE REVIEW

**Comment by riverlijunjie** [comment]
`/azp run openvino-win`

File: samples/cpp/benchmark_app/result_dump.cpp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "result_dump.hpp"
+
+#include <nlohmann/json.hpp>
+
+#include "samples/slog.hpp"
+
+void ResultDump::add_result(const std::string& input_image, const ov::TensorVector& outputs) {
```
Comment (riverlijunjie) [inline samples/cpp/benchmark_app/result_dump.cpp:11]: compare_and_save_result() will be more readable?

File: samples/cpp/benchmark_app/result_dump.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                if (0 != std::memcmp(new_output, old_output, buffsize)) {
+                    same = false;
+                    break;
+                }
+            }
+            offset += buffsize;
+            old_output += buffsize;
+        }
+
+        if (same) {
```
Comment (riverlijunjie) [inline samples/cpp/benchmark_app/result_dump.cpp:63]: Can we merge variable "same" and "different" to be one?

File: samples/cpp/benchmark_app/benchmark_app.hpp [hunk 3]
```diff
@@ -363,6 +372,15 @@ DEFINE_string(imean, "", input_image_mean_message);
 /// @brief Define flag for inference only mode <br>
 DEFINE_bool(inference_only, true, inference_only_message);
 
+/// @brief Define flag for dumping output directory <br>
+DEFINE_string(dump_output, "", dump_output_message);
+
+/// @brief Define parameter for the maximum size of dump binary file <br>
+DEFINE_uint32(dump_output_binary_size, 1024, dump_output_binary_size_message);
```
[the remark this answers is not in the document]
Comment (xufang-lisa) [reply samples/cpp/benchmark_app/benchmark_app.hpp:379]: ok, I will modify it.

File: samples/cpp/benchmark_app/result_dump.cpp (hunk 1 above)
Comment (xufang-lisa) [reply samples/cpp/benchmark_app/result_dump.cpp:11]: ok, I will modify it.

File: samples/cpp/benchmark_app/result_dump.cpp (hunk 2 above)
Comment (xufang-lisa) [reply samples/cpp/benchmark_app/result_dump.cpp:63]: I will replace "break" with "return" and delete the variable "different".

**Review by wangleis — APPROVED** [review]

**Review by riverlijunjie — APPROVED** [review]

File: samples/cpp/benchmark_app/infer_request_wrap.hpp [hunk 4]
```diff
@@ -42,7 +54,16 @@ class InferReqWrap final {
...
-            _callbackQueue(_id, _lat_group_id, get_execution_time_in_milliseconds(), ptr);
+            const auto& compile_model = _request.get_compiled_model();
+            const auto& outputs = compile_model.outputs();
+            auto output_size = outputs.size();
+            InferenceResult inference_result;
+            for (size_t i = 0; i < output_size; i++) {
+                const auto& tensor = _request.get_output_tensor(i);
+                inference_result.output_tensors.emplace_back(tensor);
+            }
+            inference_result.input_images = get_input_image_name();
```
Comment (nkogteva) [inline samples/cpp/benchmark_app/infer_request_wrap.hpp:65]: Calling this code on each callback can lead to performance degradation even for cases when dumping output data is not required.

File: samples/cpp/benchmark_app/infer_request_wrap.hpp [hunk 5]
```diff
@@ -150,11 +190,15 @@ class InferRequestsQueue final {
...
                           size_t lat_group_id,
                           const double latency,
+                          const InferenceResult& result,
                           const std::exception_ptr& ptr = nullptr) {
         std::unique_lock<std::mutex> lock(_mutex);
         if (ptr) {
             inferenceException = ptr;
         } else {
+            if (_result_dump) {
+                _result_dump->compare_and_save_result(result.input_images, result.output_tensors);
```
Comment (nkogteva) [inline samples/cpp/benchmark_app/infer_request_wrap.hpp:200]: Dumping results in parallel with inference on some lightweight topologies can be a bottleneck. Not sure that you will get what you want in this case. Requests will stuck in queue due to saving them.

**Review by nkogteva — CHANGES_REQUESTED** [review]
Not sure that benchmark_app is a right place for this functionality. We need to have a dedicated meeting to discuss all requirements.

File: samples/cpp/benchmark_app/infer_request_wrap.hpp (hunk 4 above)
Comment (xufang-lisa) [reply samples/cpp/benchmark_app/infer_request_wrap.hpp:65]: I will add conditional judgement to reduce the impact on performance.

File: samples/cpp/benchmark_app/infer_request_wrap.hpp (hunk 5 above)
Comment (xufang-lisa) [reply samples/cpp/benchmark_app/infer_request_wrap.hpp:200]: Dumping results and performance test should not be done at the same time. I will test how much the result dump affects the performance.

**Comment by akladiev** [comment]
This PR will be closed in 2 weeks in case of no activity.

**Pull request closed without merging**